core/src/main/java/io/questdb/griffin/RecordToRowCopierUtils.java

```diff
@@ -0,0 +1,1025 @@
+                                asm.ldc(toColumnType_0 + i * 2);
+                                asm.invokeStatic(implicitCastGeoHashAsGeoHash);
+                                asm.l2i();
+                            }
+                            asm.invokeInterface(wPutInt, 2);
+                            break;
+                    }
+                    break;
+                case ColumnType.GEOLONG:
+                    asm.invokeInterface(rGetGeoLong, 1);
+                    switch (ColumnType.tagOf(toColumnType)) {
+                        case ColumnType.GEOBYTE:
+                            asm.ldc(fromColumnType_0 + i * 2);
+                            asm.ldc(toColumnType_0 + i * 2);
+                            asm.invokeStatic(implicitCastGeoHashAsGeoHash);
+                            asm.l2i();
+                            asm.i2b();
+                            asm.invokeInterface(wPutByte, 2);
+                            break;
+                        case ColumnType.GEOSHORT:
+                            asm.ldc(fromColumnType_0 + i * 2);
+                            asm.ldc(toColumnType_0 + i * 2);
+                            asm.invokeStatic(implicitCastGeoHashAsGeoHash);
+                            asm.l2i();
+                            asm.i2s();
+                            asm.invokeInterface(wPutShort, 2);
+                            break;
+                        case ColumnType.GEOINT:
+                            asm.ldc(fromColumnType_0 + i * 2);
+                            asm.ldc(toColumnType_0 + i * 2);
+                            asm.invokeStatic(implicitCastGeoHashAsGeoHash);
+                            asm.l2i();
+                            asm.invokeInterface(wPutInt, 2);
+                            break;
+                        default:
+                            if (fromColumnType != toColumnType && fromColumnType != ColumnType.NULL && fromColumnType != ColumnType.GEOLONG) {
+                                asm.ldc(fromColumnType_0 + i * 2);
+                                asm.ldc(toColumnType_0 + i * 2);
+                                asm.invokeStatic(implicitCastGeoHashAsGeoHash);
+                            }
+                            asm.invokeInterface(wPutLong, 3);
+                            break;
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        asm.return_();
+        asm.endMethodCode();
+
+        // exceptions
+        asm.putShort(0);
+
+        // we have to add stack map table as branch target
+        // jvm requires it
+
+        // attributes: 0 (void, no stack verification)
+        asm.putShort(0);
+
+        asm.endMethod();
+
+        // class attribute count
+        asm.putShort(0);
+
+        return asm.newInstance();
+    }
+
+    // used by copier
+    @SuppressWarnings("unused")
+    static void checkDoubleBounds(
```

Comment:
shouldn't we also check for `NaN` here? (+in the float version of this method) 

right now, we don't do the check thus we are at [mercy of Java casting](https://github.com/questdb/questdb/blob/vi_bad_cast/core/src/main/java/io/questdb/griffin/RecordToRowCopierUtils.java#L558)

Java casts `NaN` as `0`:
```
jshell
|  Welcome to JShell -- Version 11.0.16
|  For an introduction type: /help intro

jshell> System.out.println(Double.NaN);
NaN

jshell> System.out.println((int)Double.NaN);
0
```

This differs from our [explicit casting](https://github.com/jerrinot/questdb/blob/826f3e677fd976d686cbdfa9fb3091fc58e6936d/core/src/main/java/io/questdb/griffin/engine/functions/cast/CastDoubleToIntFunctionFactory.java#L71) which treats `NaN` as `null`

This means that given this table:
```sql
CREATE TABLE convert_demo (
    col_int int
);
```
then:
* Explicit casting `insert into 'convert_demo' values (cast (cast ('NaN' as float) as int));` insert `null` into the table
* Implicit casting `insert into 'convert_demo' values (cast ('NaN' as float));` insert number `0`.

Reply:
fwiw, Postgres does not allow casting `NaN` to `int` at all, even explicit casting fails:
this returns a float column: `select cast(cast ('NaN' as float) as int);`
but this fails `select cast(cast ('NaN' as float) as int);` with `[22003] ERROR: integer out of range`

Reply:
fixed in https://github.com/questdb/questdb/pull/2423/commits/3202b74e869e4b25fc7eea0bafef99749f0fe617

although our `int` does support `null` therefore we can cast double & float NaNs to `int` and `long`. Casting `NaN` to smaller types yields 0